pkg/route/error.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package route
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/errutil"
+)
+
+var (
+	// errConflict is returned whenever an operation fails due to any conflict-type error.
+	errConflict = errors.New("conflict")
+
+	errBadReplica  = errors.New("replica count exceeds replication factor")
+	errNotReady    = errors.New("target not ready")
+	errUnavailable = errors.New("target not available")
+)
+
+type expectedErrors []*struct {
+	err   error
+	cause func(error) bool
+	count int
+}
+
+func (a expectedErrors) Len() int           { return len(a) }
+func (a expectedErrors) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a expectedErrors) Less(i, j int) bool { return a[i].count < a[j].count }
+
+// determineWriteErrorCause extracts a sentinel error that has occurred more than the given threshold from a given fan-out error.
+// It will inspect the error's cause if the error is a MultiError,
+// It will return cause of each contained error but will not traverse any deeper.
+func determineWriteErrorCause(err error, threshold int) error {
```

Comment:
Let's make sure we have parity https://github.com/thanos-io/thanos/blob/befb0257e8a3d49cb80b00794a4892abfb7ccb72/pkg/receive/handler.go#L689

Reply:
Can you elaborate on this?